ivadomed/postprocessing.py

```diff
@@ -0,0 +1,205 @@
+# Deals with postprocessing on generated segmentation.
+
+import numpy as np
+import nibabel as nib
+from scipy.ndimage.measurements import label
+from scipy.ndimage.morphology import binary_fill_holes
+
+
+def threshold_predictions(predictions, thr=0.5):
+    """Threshold predictions.
+
+    Threshold a soft (ie not binary) array of predictions given a threshold value, and returns
+    a binary array.
+
+    Args:
+        predictions (array): array to binarise.
+        thr (float): Threshold value: voxels with a value < to thr are assigned 0 as value, 1
+            otherwise.
+    Returns:
+        array: Array containing only zeros or ones.
+
+    """
+    thresholded_preds = np.copy(predictions)[:]
+    low_values_indices = thresholded_preds < thr
+    thresholded_preds[low_values_indices] = 0
+    low_values_indices = thresholded_preds >= thr
+    thresholded_preds[low_values_indices] = 1
+    return thresholded_preds.astype(np.int)
+
+
+def threshold_predictions_nib(nib_predictions, thr=0.5):
+    """Threshold predictions from a nibabel object.
+
+    Threshold a soft (ie not binary) array of predictions from a nibabel object given a threshold value,
+    and returns a nibabel object with binary data.
+    Note: this function calls threshold_predictions().
+
+    Args:
+        nib_predictions (nibabelObject): nibabel image to binarise.
+        thr (float): Threshold value: voxels with a value < to thr are assigned 0 as value, 1
+            otherwise.
+    Returns:
+        nibabelObject: Nibabel object containing only zeros or ones.
+
+    """
+    data = nib_predictions.get_data()
+    data_thr = threshold_predictions(predictions=data, thr=thr)
+    return nib.Nifti1Image(data_thr, nib_predictions.affine)
+
+
+def keep_largest_object(predictions):
+    """Keep the largest connect object.
+
+    Keep the largest connected object from the input array (2 or 3D).
+    Note: This function only works for binary segmentation.
+
+    Args:
+        predictions (array): Input 2 or 3D binary segmentation.
+    Returns:
+        array: processed segmentation.
+
+    """
+    assert predictions.dtype == np.dtype('int')
+    # Find number of closed objects using skimage "label"
+    labeled_obj, num_obj = label(predictions)
+    # If more than one object is found, keep the largest one
+    if num_obj > 1:
+        # Keep the largest object
+            predictions[np.where(labeled_obj != (np.bincount(labeled_obj.flat)[1:].argmax() + 1))] = 0
+    return predictions
+
+
+def keep_largest_object_nib(nib_predictions):
```

Comment:
i know we've discussed this and i suggested to split this up in two functions (one for nib, one for np.array), but we could save some lines of code if we accommodate both types within the same functions, and deal with i/o at the beginning/end of the function, as done [here](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/resampling.py#L50) for example

Reply:
in fact, instead of duplicating the "look for type, if nibabel do X, then do Y", which is the same for most postprocessing functions, i am wondering if there is a python trick to do this more elegantly. @kousu any suggestion?

Reply:
The basic issue with all these `op(im)` vs `op_nib(im)` cases is that you want the types to be line up:

* `op: np.array => np.array`
* `op_nib: nib.Nifti1Image => nib.Nifti1Image`

If you just wanted the be able to handle both on the input side, you could decide

* `op: (np.array | nib.Nifti1Image) => nib.Nifti1Image`

and handle them with a typecast a the top of the function, so:

```
def keep_largest_object(predictions):
    """Keep the largest connect object.
    Keep the largest connected object from the input array (2 or 3D).
    Note: This function only works for binary segmentation.
    Args:
        predictions (array): Input 2 or 3D binary segmentation.
    Returns:
        array: processed segmentation.
    """
    predictions = nib.Nifti1Image(predictions)
    assert predictions.get_fdata().dtype == np.dtype('int')
    # Find number of closed objects using skimage "label"
    labeled_obj, num_obj = label(predictions.get_fdata())
    # If more than one object is found, keep the largest one
    if num_obj > 1:
        # Keep the largest object
            predictions.get_fdata()[np.where(labeled_obj != (np.bincount(labeled_obj.flat)[1:].argmax() + 1))] = 0
    return nib.Nifti1Image(predictions.get_fdata(), predictions.affine)
```

This requires `nib.Nifti1Image()` to be able to handle taking a plain `np.array` alone sensibly, and to handle being passed another `nib.Nifti1Image` as a copy.

But if you want the data to come out the same as the input type then I think as written is as good as you're going to get. The one thing you could maybe do is merge them into one function with a recursion:

```
def keep_largest_object(predictions):
    """Keep the largest connect object.
    Keep the largest connected object from the input array (2 or 3D).
    Note: This function only works for binary segmentation.
    Args:
        predictions (array): Input 2 or 3D binary segmentation.
    Returns:
        array: processed segmentation.
    """
    if isinstance(predictions, nib.Nifti1Image):
        data_out = keep_largest_object(predictions.get_fdata())
        return nib.Nifti1Image(data_out, predictions.affine)
        
    assert predictions.dtype == np.dtype('int')
    # Find number of closed objects using skimage "label"
    labeled_obj, num_obj = label(predictions)
    # If more than one object is found, keep the largest one
    if num_obj > 1:
        # Keep the largest object
            predictions[np.where(labeled_obj != (np.bincount(labeled_obj.flat)[1:].argmax() + 1))] = 0
    return predictions
```

Reply:
>  The one thing you could maybe do is merge them into one function with a recursion:

Love it! i'll refactor according to this suggestion

Reply:
Great! In general one can emulate Julia/C++/Java's idea of type-based function overloading with recursions, but it's definitely a bit more verbose in python.

Reply:
There actually might be a "python trick" to making this easier too. We could write a decorator:

```
def nib(wrapped):
    @functools.wraps(wrapped)
    def wrapper(data, *args, **kwargs):
        if isinstance(data, nib.NiftiImage):
            return nib.Nifti1Image(f(data.dataobj.get_unscaled()), data.affine)
        return f(data)
    return wrapper

...

@nib
def keep_largest_object(predictions):
   ....

@nib
def fill_holes(predictions):
   ...
```

Reply:
omg this is beautiful! love it! thanks @kousu !!